### PR TITLE
#87 TEPS 2023

### LINKED ISSUE #87: TEPS 2023
Extract 2023 data for Toronto Emission Prediction Scheme (TEPS)
- Miovision ATR data to the centreline
- ATR counts from MOVE to the centreline
- Vehicle Detector Stations (to the centreline)

See past issue: #80

### PR DESCRIPTION
**Pull request by gabrielwol**
**Summary of edits:**
- New Schema: `teps`
- Did not reexamine the RESCU thresholds.
- We are reusing the 2022 centreline + conflation.
- `volume_project/extract/2022/miovision_sqls/miovision_centreline_20220705.sql`: Sare, I think you already fixed most of these duplicate issues with the miovision-centreline but had not committed them so I had to recreate them. Will use this as the basis for [#923](https://github.com/CityofToronto/bdit_data-sources/issues/923).
- `volume_project/extract/2022/rescu_sqls/rescu_centreline.sql`: Added the 2023 rescu detectors to the centreline conflation script. There were 4 nulls that I manually assigned + an update script for the missing geoms. 
- `volume_project/extract/2023/miovision_sqls/miovision_atr_2023.sql`: just a rename.
- The rest is just fluff!

**Some high level checks:** 
**RESCU:** Despite the outages we have noticed the last few months of 2023, there was more data for in 2023 than 2022.
```sql
SELECT yr, COUNT(*) AS num_detectors, ROUND(AVG(count), 1) AS avg_days, SUM(count) AS sensor_x_days
FROM (
    SELECT detector_id, '2022' AS yr, COUNT(*) FROM teps.rescu_enuf_vol_22 GROUP BY 1
    UNION
    SELECT detector_id, '2023' AS yr, COUNT(*) FROM teps.rescu_enuf_vol_23 GROUP BY 1
) AS summ
GROUP BY yr
```
| "yr"   | "num_detectors" | "avg_days" | "sensor_x_days" |
|--------|-----------------|------------|-----------------|
| "2022" | 98              | 144.1      | 14125           |
| "2023" | 98              | 205.4      | 20131           |

**Classification counts:** 
- There was [some concern](https://bditto.slack.com/archives/C06TQ1Z4D8X/p1714416422502659) that Spectrum classification counts have not started making it into the database yet, which shows up in the max date of 2023-08 below. However, there are still more counts conducted in 2023 vs 2022 overall.
```sql
SELECT '2022' AS yr, MIN(count_bin), MAX(count_bin), COUNT(DISTINCT volume_id) FROM teps.classvol_20220705
UNION
SELECT '2023' AS yr, MIN(count_bin), MAX(count_bin), COUNT(DISTINCT volume_id) FROM teps.classvol_2023;
```
| "yr"   | "min"                 | "max"                 | "count" |
|--------|-----------------------|-----------------------|---------|
| "2022" | "2022-03-21 00:00:00" | "2022-11-24 23:45:00" | 234     |
| "2023" | "2023-03-28 00:00:00" | "2023-08-03 23:45:00" | 807     |

Results look good for `speedvol` tables: 
| "yr"   | "min"                 | "max"                 | "count" |
|--------|-----------------------|-----------------------|---------|
| "2022" | "2022-02-22 00:00:00" | "2022-12-05 23:45:00" | 5665    |
| "2023" | "2023-01-10 00:00:00" | "2023-12-07 23:45:00" | 5027    |

**Miovision - centreline check:**
Double checked and there were 4 legs missing from miovision centreline. All are valid omissions: 
```sql
/*
5	"W" --bathurst and front, 3 legged intersection, delete this approach
78	"S" --private entrance at bloor and kingsway, delete this approach
68	"N" --this approach is north of steeles so outside of Toronto centreline.
1	"W" --bathurst and adelaide, 3 legged intersection, delete this approach.
*/
WITH mio_data AS (
    SELECT DISTINCT
        intersection_uid, leg
    FROM miovision_api.volumes_15min AS volumes
    WHERE
        volumes.datetime_bin >= '2023-01-01 00:00:00'::timestamp without time zone
        AND volumes.datetime_bin < '2024-01-01 00:00:00'::timestamp without time zone 
        AND (volumes.classification_uid <> ALL (ARRAY[2, 6, 10]))
)

SELECT mio_data.*, cm.centreline_id FROM mio_data
LEFT JOIN teps.centreline_miovision_20220705 cm
    ON mio_data.intersection_uid = cm.intersection_uid
    AND mio_data.leg = cm.leg
WHERE cm.centreline_id IS NULL
```

### CODE REVIEW

File: volume_project/extract/2023/short_term_count_sqls/tmc_2_atr_vol.sql [hunk 1]
```diff
@@ -1,62 +1,62 @@
...
+            mlf.arterycode,
+            mlf.traffic_column_name,
+            mlf.datetime_bin,
+            mlf.volume,
+            mlf.traffic_classification,
+            tt.tcl_from_segment,
+            tt.tcl_to_segment,
+            tt.from_dir,
+            tt.to_dir
+        FROM traffic.tmc_miovision_long_format AS mlf
```
Comment (scann0n) [inline volume_project/extract/2023/short_term_count_sqls/tmc_2_atr_vol.sql:18]: `traffic.det` contains TMC data to the end of 2023 - `traffic.tmc_miovision_long_format` should be updated to include this info, then this mat view should be updated to include the September to December data.

File: volume_project/extract/2023/short_term_count_sqls/tmc_2_atr_check.sql [hunk 2]
```diff
@@ -7,121 +7,122 @@
...
 
 -- Blythwood and Strathgowan
 SELECT centreline_id, lf_name, fcode_desc, sum(volume)
-FROM teps.tmcs_to_atrs_20220705
+FROM teps.tmcs_to_atrs_2023
 LEFT JOIN gis.centreline_20220705 ON centreline_id = geo_id
 WHERE centreline_id IN (7754013, 30057003, 1137956)
-GROUP BY centreline_id, lf_name, fcode_desc
+GROUP BY centreline_id, lf_name, fcode_desc;
 -- These all checked out so I stopped.
```
Comment (scann0n) [inline volume_project/extract/2023/short_term_count_sqls/tmc_2_atr_check.sql]: Consider checking three other intersections (not a show stopper)

**Review by scann0n — COMMENTED** [review]
Please see the comments re: updating the TMCs to ATRs table (which requires an update of the long form table). I think we can give them more data...

I checked the geo conflagrations for miovision and rescu and everything looks great!!!

File: volume_project/extract/2023/short_term_count_sqls/tmc_2_atr_vol.sql (hunk 1 above)
Comment (gabrielwol) [reply volume_project/extract/2023/short_term_count_sqls/tmc_2_atr_vol.sql:18]: Good catch! Running: 

```sql
REFRESH MATERIALIZED VIEW traffic.tmc_miovision_long_format WITH DATA;
REFRESH MATERIALIZED VIEW teps.tmcs_to_atrs_2023 WITH DATA;
```

File: volume_project/extract/2023/short_term_count_sqls/tmc_2_atr_check.sql (hunk 2 above)
Comment (gabrielwol) [reply volume_project/extract/2023/short_term_count_sqls/tmc_2_atr_check.sql]: @scann0n check out the latest checks!

**Review by scann0n — APPROVED** [review]
Amazing work!!! I love that we're using data dumps like this to DQ our data!!!

**Pull request merged**